### DISCUSSION TITLE
Multi platform docker build - linux/mac/windows

### DISCUSSION THREAD

**Discussion by grzesuav** [goreleaser/goreleaser#3314] [Q&A] [answered]
Hi,

I have successfully set up `goreelaser` flow - https://github.com/metacontroller/metacontroller/releases/tag/v4.3.5 where I build binaries for various arch, and then docker images for linux-based arch's. I would like additionally to build macos/windows images, however I need to do that in different job (on different runner).

I was thinking of :
* passing build artifacts, which are already created
* use gorelease to just do `docker` and `docker-manifest` step, which will build images for missing architectures and add them (via `manifest`) to existing multi archi images

would it be possible ? What would be the easiest way to achieve that ?

Thanks for the great tool !

**Comment by caarlos0** [comment]
I'm not aware of anyone doing that... but let us know what you try and what works 🙏

**Reply by grzesuav** [reply]
@caarlos0 I can imagine I can spin another worker and pass artifacts, how can I invoke goreleaser to perform just singe docker build job?

I wonder if it can update `docker-manifest` from previous run or it will override it ?

**Comment by 6543** [comment]
We use docker buildx https://github.com/woodpecker-ci/woodpecker/blob/master/.woodpecker/docker.yml#L69

Works also outside of ci ...

**Reply by grzesuav** [reply]
@6543 it works for all *nix, I do not see windows or macos platform in buildx on my linux machine, I also do not see it in your config, did you even try to build macos or windows image ?

**Comment by caarlos0** [comment, marked as answer]
for the particular case of windows, you might need to use goreleaser pro with split/merge pipelines: https://carlosbecker.com/posts/goreleaser-split-merge/

for linux only, `docker buildx` should work just fine.

**Reply by mdelapenya** [reply]
From your comment I understand that building Windows containers with goreleaser is a Pro feature, isn't it?

I'm trying to get it working in [this branch](https://github.com/mdelapenya/moby-ryuk/tree/wcow-goreleaser) without success:

```
  ⨯ release failed after 8s                  error=docker build failed: failed to build docker.io/testcontainers/ryuk:0.3.4-SNAPSHOT-2b66401-win1809-amd64: exit status 1: #1 [internal] load build definition from Dockerfile
#1 transferring dockerfile:
#1 transferring dockerfile: 174B done
#1 DONE 0.0s

#2 [internal] load .dockerignore
#2 transferring context: 2B done
#2 DONE 0.0s

#3 [internal] load metadata for mcr.microsoft.com/windows/nanoserver:1809
#3 DONE 0.4s

#4 [1/2] FROM mcr.microsoft.com/windows/nanoserver:1809@sha256:5c9da3bfe1524eab610dfd41c79a0805c080f8307d810d62bba106b5df587698
#4 DONE 0.0s

#5 [internal] load build context
#5 transferring context: 6.81MB 0.1s done
#5 DONE 0.1s

#4 [1/2] FROM mcr.microsoft.com/windows/nanoserver:1809@sha256:5c9da3bfe1524eab610dfd41c79a0805c080f8307d810d62bba106b5df587698
#4 CACHED

#6 [2/2] COPY ryuk.exe /
#6 DONE 0.0s

#7 exporting to image
#7 exporting layers 0.0s done
#7 writing image sha256:9b67fefc00c3ed848d94f0d3b318549d6d55b043edc94200cb9a4aed4dcc3c2a done
#7 ERROR: operating system is not supported
------
 > exporting to image:
------
ERROR: failed to solve: operating system is not supported

Learn more at https://goreleaser.com/errors/docker-build
```

**Reply by caarlos0** [reply]
That's a Docker error... I _think_  it's only possible to build windows images on a windows box, will try it later once I'm at a keyboard

**Reply by mdelapenya** [reply]
I had the same feeling, and tried in a Windows 11 Pro with the following error:

> ERROR: failed to solve: rpc error: code = Unknown desc = failed to solve with frontend dockerfile.v0: failed to read dockerfile: failed to mount in-place: {bind C:\ProgramData\Docker\windowsfilter\rn22w4gy5o12y1n0ta394rjdf [rbind ro]}: invalid windows mount type: 'bind'

We checked a few things in that box: Windows containers were enabled in the DockerForWin app, disabling buildx...

I also received this error on my Mac M1:

> ERROR: failed to solve: rpc error: code = Unknown desc = failed to solve with frontend dockerfile.v0: failed to create LLB definition: no match for platform in manifest sha256:5c9da3bfe1524eab610dfd41c79a0805c080f8307d810d62bba106b5df587698: not found

**Reply by mdelapenya** [reply]
BTW thanks for the fast feedback, much appreciated 🙏

**Reply by caarlos0** [reply]
Okay, confirmed: you can't build windows images on linux hosts.

That said, I created another example showing how to do it with goreleaser pro + split & merge builds.

There was already an example for split and merge stuff, but not for the windows docker image part.

- You can read a bit about it [here](https://carlosbecker.com/posts/goreleaser-split-merge/)
- The example for windows docker images: https://github.com/caarlos0/goreleaser-pro-split-merge-example-real
- The example of split/merge without windows stuff: https://github.com/caarlos0/goreleaser-pro-split-merge-example
- A release build output: https://github.com/caarlos0/goreleaser-pro-split-merge-example-real/actions/runs/3588449110
- and the manifest: https://github.com/caarlos0/goreleaser-pro-split-merge-example-real/pkgs/container/goreleaser-pro-split-merge-example-real/55803713?tag=latest

Hopefully that's helpful :)

**Reply by caarlos0** [reply]
PS: there are only a few changes from the no-windows example to the other, you can copy one project over the other and `git diff` to see them all :)

**Reply by grzesuav** [reply]
yeah, I would need this in open-source project, so using a Pro version is kind of no-go for me

**Reply by mdelapenya** [reply]
@caarlos0 Do you think this way to parallel cross-build Linux and Windows would work? https://github.com/testcontainers/moby-ryuk/pull/52

**Reply by caarlos0** [reply]
on a quick glance, it seems that both runs will write the same docker manifest, so one will override the other *I think*...

**Reply by mdelapenya** [reply]
Thanks for taking a look 🙏 I wonder if both parallel branches will annotate the already existing docker manifest, created in the stage that is previous to the parallel stage 🤔

**Reply by caarlos0** [reply]
I think it will override it, but it might be worth taking a look and fixing it if that's the case

**Reply by mdelapenya** [reply]
In any case, thanks for your guidance here, and for the goreleaser project 🚀
